### ISSUE TITLE
Low F1 score for the miscellaneous class

### ISSUE THREAD

**Issue by GekFreeman** [vqdang/hover_net#28]
Thanks for your excellent work first!
I got a result on the CoNSeP dataset as follows:
| Metrics | F<sup>m</sup>|
| ------- | ---------------- |
|   Mine   | 0.042 |
|In Paper| 0.426  |

I find that the F1 score for the miscellaneous class is significantly lower than the record in your paper. I have tried a lot to find out the reasons but failed...Could you help me with this problem?
P.S. I should expect a lower F1 score for the miscellaneous class because there are significantly fewer nuclei represented, but it seems something is wrong...

Some other results
|Metrics|F<sup>i</sup>|F<sup>e</sup>|F<sup>s</sup>|
|-|-|-|-|
|Mine|0.556|0.628|0.532|
|In Paper|0.631|0.635|0.566|

**Comment by vqdang** [comment]
@GekFreeman Hi, have you converted the class within CoNSeP to correct label? This is my main suspect. 
If everything is right you should have results close to the paper. Case in point, if you check https://github.com/vqdang/hover_net/issues/16#issuecomment-573297705 edited history, you can see the screenshot of his experimental runs, which are even slightly better than our paper version.

**Comment by vqdang** [comment]
You have to make sure that these portions are un-commented if you are using CoNSeP. https://github.com/vqdang/hover_net/blob/eaacccec9bb3051eab87d2805e625befda949ba8/src/compute_stats.py#L55-L58

**Comment by simongraham** [comment]
Hi @GekFreeman 

Can you please show us a screen shot of the valid dice score for the miscellaneous class from tensorboard please.

**Comment by GekFreeman** [comment]
Hi @vqdang @simongraham 
I did uncomment the converting code, and here is the screenshot of the valid dice score for the miscellaneous class
![image](https://user-images.githubusercontent.com/12059962/76155298-4a82ed00-6125-11ea-96fa-2a3771c06ac5.png)
And following your guide, I find the valid dice scores for the other classes are always 0, like
![image](https://user-images.githubusercontent.com/12059962/76155331-b82f1900-6125-11ea-89ec-26431ed1f484.png)
Is that correct?

**Comment by GekFreeman** [comment]
I run the infer.py with the whole images (test set), instead of using the patches, is this the problem?

**Comment by vqdang** [comment]
These don't look right. I think something is botched during your training. Can you attach the log.log within folder 00 and 01 here?

**Comment by GekFreeman** [comment]
Here are my logs:
[log_00.log](https://github.com/vqdang/hover_net/files/4305139/log_00.log)
[log_01.log](https://github.com/vqdang/hover_net/files/4305141/log_01.log)

And here is the main code where I have modified in config.py
```python 
        self.seed = 10 
        mode = 'hover'
        self.model_type = 'np_hv'

        self.type_classification = True
        self.nr_types = 5  # denotes number of classes for nuclear type classification
        # ! some semantic segmentation network like micronet,
        # ! nr_types will replace nr_classes if type_classification=True
        self.nr_classes = 2 # Nuclei Pixels vs Background

        #### Dynamically setting the config file into variable
        if mode == 'hover':
            config_file = importlib.import_module('opt.hover') # np_hv, np_dist
        else:
            config_file = importlib.import_module('opt.other') # fcn8, dcan, etc.
        config_dict = config_file.__getattribute__(self.model_type)

        for variable, value in config_dict.items():
            self.__setattr__(variable, value)
        #### Training data

        # patches are stored as numpy arrays with N channels
        # ordering as [Image][Nuclei Pixels][Nuclei Type][Additional Map]
        # Ex: with type_classification=True
        #     HoVer-Net: RGB - Nuclei Pixels - Type Map - Horizontal and Vertical Map
        # Ex: with type_classification=False
        #     Dist     : RGB - Nuclei Pixels - Distance Map
        data_code_dict = {
            'unet'     : '536x536_84x84',
            'dist'     : '536x536_84x84',
            'fcn8'     : '512x512_256x256',
            'dcan'     : '512x512_256x256',
            'segnet'   : '512x512_256x256',
            'micronet' : '504x504_252x252', 
            'np_hv'    : '540x540_80x80',
            'np_dist'  : '540x540_80x80',
        }

        self.data_ext = '.npy' 
        # list of directories containing validation patches
        self.train_dir = ['../dataset/CoN-train/%s/Train/XXXX/' % data_code_dict[self.model_type]]
        self.valid_dir = ['../dataset/CoN-train/%s/Train/XXXX/' % data_code_dict[self.model_type]]

        # nr of processes for parallel processing input
        self.nr_procs_train = 8  
        self.nr_procs_valid = 4  

        self.input_norm  = True # normalize RGB to 0-1 range

        ####
        #
        exp_id = 'v1.0.1.6/CoNSeP'
        model_id = '%s' % self.model_type
        self.model_name = '%s/%s' % (exp_id, model_id)
        # loading chkpts in tensorflow, the path must not contain extra '/'
        self.log_path = '../experiment' # log root path
        self.save_dir = '%s/%s' % (self.log_path, self.model_name) # log file destination

        #### Info for running inference
        self.inf_auto_find_chkpt = True
        # path to checkpoints will be used for inference, replace accordingly
        self.inf_model_path  = self.save_dir + '/01/model-57000.index'

        # output will have channel ordering as [Nuclei Type][Nuclei Pixels][Additional]
        # where [Nuclei Type] will be used for getting the type of each instance
        # while [Nuclei Pixels][Additional] will be used for extracting instances

        # TODO: encode the file extension for each folder?
        # list of [[root_dir1, codeX, subdirA, subdirB], [root_dir2, codeY, subdirC, subdirD] etc.]
        # code is used together with 'inf_output_dir' to make output dir for each set
        self.inf_imgs_ext = '.png'
        self.inf_data_list = [
            ['../dataset/CoNSeP/Test', 'XXXX', 'Images'],
        ]
        self.inf_output_dir    = 'output/%s/%s' % (exp_id, model_id)

        # for inference during evalutaion mode i.e run by inferer.py
        self.eval_inf_input_tensor_names = ['images']
        self.eval_inf_output_tensor_names = ['predmap-coded']
        # for inference during training mode i.e run by trainer.py
        self.train_inf_output_tensor_names = ['predmap-coded', 'truemap-coded']
```

Thanks a lot!

**Comment by vqdang** [comment]
@GekFreeman Hi, I have checked your `log.log`, the content is very wrong. For comparison, this is the content of a training run I did last Sunday to check everything again, using a new conda environment and latest code from github with a 32GB V100.

[log_00.log](https://github.com/vqdang/hover_net/files/4306873/log.log)
[log_01.log](https://github.com/vqdang/hover_net/files/4306877/log.log)

Within your `log.log`
```
[32m[0304 20:25:08 @monitor.py:459][0m valid_acc: 0.8582
[32m[0304 20:25:08 @monitor.py:459][0m valid_dice: 0.25335
[32m[0304 20:25:08 @monitor.py:459][0m valid_dice_epi: 0
[32m[0304 20:25:08 @monitor.py:459][0m valid_dice_fib: 0
[32m[0304 20:25:08 @monitor.py:459][0m valid_dice_inf: 0
[32m[0304 20:25:08 @monitor.py:459][0m valid_dice_msc: 0.10004
[32m[0304 20:25:08 @monitor.py:459][0m valid_mse: 0.051064
```
This indicates something is broken within the validation calculation. Also, from the code pieces that you attached, `self.valid_dir` points to the same dir as `self.train_dir` so you may want to fix it also. 

All in all, I don't know what was broken in your version. So to be sure, I suggest that you do the following to make the debugging easier.
1. Make sure the patches are generated for both the training set and validation set correctly.
2. Rerun the experiment using the latest version of the code available on github. 
3. Make sure that non-zero values are printed out for any validation metrics.

**Comment by GekFreeman** [comment]
I get the right valid loss with the latest code, by the way, how did you choose the validation set for CoNSeP datasets to get the result in your paper?

**Comment by simongraham** [comment]
Okay, sounds good about the model now training as expected. We do not provide an additional independent validation set and therefore you can create one from within the training set. 

It seems that the above issue is now solved and will therefore close.

**Comment by haiqinzhong** [comment]
@GekFreeman ，Hi, Can you teach me how do you get valid set from train set? I don't know whether I 'm right

**Comment by GekFreeman** [comment]
I get the valid set by:
1.extract patches from the training set
2.randomly split the patches (10k fold or other)

**Comment by haiqinzhong** [comment]
> I get the valid set by:
> 1.extract patches from the training set
> 2.randomly split the patches (10k fold or other)
@GekFreeman , thank you for your reply.
   So, firstly, you extract patches from the trainning set(27 images). And then use 10k fold to split these patches to get the valid set. Right? My way is get valid set from training set(randomly split 20%, 6 images), and then extract patches from training set and valid set, respectively. 
   And the result is very bad, how about your result, is it close to author's?
   It very important for me, I would appreciate your reply.

**Comment by GekFreeman** [comment]
I did not find the results of my experiment, so I am sorry I cannot tell you the exact results, but my performance on the test set is close to that of the author, but it is still slightly worse.

**Comment by haiqinzhong** [comment]
OK,I&#39;m try your way



------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Freeman"<notifications@github.com&gt;; 
发送时间: 2020年11月6日(星期五) 下午4:29
收件人: "vqdang/hover_net";
抄送: "Comment";
主题: Re: [vqdang/hover_net] Low F1 score for the miscellaneous class (#28)





 
I did not find the results of my experiment, so I am sorry I cannot tell you the exact results, but my performance on the test set is close to that of the author, but it is still slightly worse.
 
&mdash;
